app/scripts/modules/core/src/presentation/tables/TableRow.tsx

```diff
@@ -0,0 +1,83 @@
+import * as React from 'react';
+
+import { ITableCellPropsWithInternalFields } from './TableCell';
+import { ITableContext, TableContext } from './TableContext';
+
+const { useState, useContext, useEffect } = React;
+
+export type ITableRowLayoutProps = ITableRowProps &
+  Pick<ITableContext, 'expandable' | 'isMobile' | 'columns'> &
+  Pick<ITableRowProps, 'renderExpandedContent'> & {
+    expanded: boolean;
+    setExpanded: (expanded: boolean) => any;
+  };
+
+export interface ITableRowProps {
+  defaultExpanded?: boolean;
+  renderExpandedContent?: () => React.ReactNode;
+  children?: React.ReactNode;
+}
+
+const logDefaultExpandedError = () => {
+  console.error(
+    new Error(
+      `You used the defaultExpanded prop on a row inside a <Table/> that's not expandable.
+       Either remove the defaultExpanded prop or make the <Table/> expandable.`,
+    ),
+  );
+};
+
+const logExpandedContentError = () => {
+  console.error(new Error(`Rows inside an expandable <Table/> must provide a renderExpandedContent prop.`));
+};
+
+const throwCellColumnError = () => {
+  throw new Error(
+    `A <TableCell/> was added that didn't match any of the columns defined on its parent <Table/>.
+     Check if there are an equal number of <TableCell/> elements to the number of columns provided to the <Table/>.
+    `,
+  );
+};
+
+export const TableRow = ({ defaultExpanded, renderExpandedContent, children }: ITableRowProps) => {
+  const { layout, expandable, allExpanded, setAllExpanded, isMobile, columns } = useContext(TableContext);
+  const [expanded, setExpanded] = useState(defaultExpanded || false);
+
+  // Expand individual rows when the table expands all its rows at once
+  useEffect(() => {
+    allExpanded && setExpanded(true);
+  }, [allExpanded]);
+
+  if (!expandable && defaultExpanded != null) {
+    logDefaultExpandedError();
+  }
+
+  if (expandable && !renderExpandedContent) {
+    logExpandedContentError();
+  }
+
+  const childrenWithIndex = React.Children.map(children, (cell, index) => {
+    return React.cloneElement(cell as React.ReactElement<ITableCellPropsWithInternalFields>, { index });
+  });
```

Comment:
not thrilled with having to do a children map+clone here, but it's in service of being able to give an individual `TableCell` the name of the column it represents so folks don't have to manually wire up names across many different components. I thought that was a pretty valuable thing to add, so I took the tradeoff and added some validation below to alert folks to potential misconfiguration (when there are the wrong number of cells compared to columns).